### PR TITLE
Adding a metrics report to IO gatherers

### PR DESCRIPTION
**Pull request by rluders**
It adds some metrics to log and to generated archive that collects some information about the gathers executions:

- Elapsed time;
- Number of records generated;
- List of errors;

The report is stored at the archive, in the file `insights-operator\gathers.json`

### CODE REVIEW

**Comment by 0sewa0** [comment]
/lgtm 

tested it locally and it produced the following metrics: (one of the entries has `"errors": []` while every other has `"errors": null`)

```
[
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherPodDisruptionBudgets.func1",
        "elapsed": 138000000,
        "report": 2,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherMostRecentMetrics.func1",
        "elapsed": 41000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterOperators.func1",
        "elapsed": 3333000000,
        "report": 48,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherContainerImages.func1",
        "elapsed": 427000000,
        "report": 12,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherNodes.func1",
        "elapsed": 585000000,
        "report": 6,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherConfigMaps.func1",
        "elapsed": 131000000,
        "report": 10,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterVersion.func1",
        "elapsed": 125000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterID.func1",
        "elapsed": 0,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterInfrastructure.func1",
        "elapsed": 131000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterNetwork.func1",
        "elapsed": 124000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterAuthentication.func1",
        "elapsed": 122000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterImageRegistry.func1",
        "elapsed": 127000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterImagePruner.func1",
        "elapsed": 127000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterFeatureGates.func1",
        "elapsed": 123000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterOAuth.func1",
        "elapsed": 181000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterIngress.func1",
        "elapsed": 123000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterProxy.func1",
        "elapsed": 121000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherCertificateSigningRequests.func1",
        "elapsed": 178000000,
        "report": 0,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherCRD.func1",
        "elapsed": 273000000,
        "report": 2,
        "errors": []
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherHostSubnet.func1",
        "elapsed": 126000000,
        "report": 6,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherMachineSet.func1",
        "elapsed": 169000000,
        "report": 3,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherInstallPlans.func1",
        "elapsed": 9693000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherServiceAccounts.func1",
        "elapsed": 10326000000,
        "report": 1,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherMachineConfigPool.func1",
        "elapsed": 217000000,
        "report": 2,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherContainerRuntimeConfig.func1",
        "elapsed": 122000000,
        "report": 0,
        "errors": null
    },
    {
        "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherStatefulSets.func1",
        "elapsed": 10274000000,
        "report": 2,
        "errors": null
    }
]
```

**Comment by tremes** [comment]
Is the duration above in ms ? It should be obvious from the file IMO. Do we need to have the name with the `github.com/openshift/insights-operator` prefix?

**Comment by rluders** [comment]
> /lgtm
> 
> tested it locally and it produced the following metrics: (one of the entries has `"errors": []` while every other has `"errors": null` which is a bit weird, but I don't think its because of this PR)
> 
> ```
> [
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherPodDisruptionBudgets.func1",
>         "elapsed": 138000000,
>         "report": 2,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherMostRecentMetrics.func1",
>         "elapsed": 41000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterOperators.func1",
>         "elapsed": 3333000000,
>         "report": 48,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherContainerImages.func1",
>         "elapsed": 427000000,
>         "report": 12,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherNodes.func1",
>         "elapsed": 585000000,
>         "report": 6,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherConfigMaps.func1",
>         "elapsed": 131000000,
>         "report": 10,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterVersion.func1",
>         "elapsed": 125000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterID.func1",
>         "elapsed": 0,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterInfrastructure.func1",
>         "elapsed": 131000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterNetwork.func1",
>         "elapsed": 124000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterAuthentication.func1",
>         "elapsed": 122000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterImageRegistry.func1",
>         "elapsed": 127000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterImagePruner.func1",
>         "elapsed": 127000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterFeatureGates.func1",
>         "elapsed": 123000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterOAuth.func1",
>         "elapsed": 181000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterIngress.func1",
>         "elapsed": 123000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherClusterProxy.func1",
>         "elapsed": 121000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherCertificateSigningRequests.func1",
>         "elapsed": 178000000,
>         "report": 0,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherCRD.func1",
>         "elapsed": 273000000,
>         "report": 2,
>         "errors": []
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherHostSubnet.func1",
>         "elapsed": 126000000,
>         "report": 6,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherMachineSet.func1",
>         "elapsed": 169000000,
>         "report": 3,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherInstallPlans.func1",
>         "elapsed": 9693000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherServiceAccounts.func1",
>         "elapsed": 10326000000,
>         "report": 1,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherMachineConfigPool.func1",
>         "elapsed": 217000000,
>         "report": 2,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherContainerRuntimeConfig.func1",
>         "elapsed": 122000000,
>         "report": 0,
>         "errors": null
>     },
>     {
>         "name": "github.com/openshift/insights-operator/pkg/gather/clusterconfig.GatherStatefulSets.func1",
>         "elapsed": 10274000000,
>         "report": 2,
>         "errors": null
>     }
> ]
> ```

Yeah, it is weird. I can omit the `errors` if it is null, but, handle it as a `[]` is also possible, just messy. I know that there is a [proposal to implement something](https://github.com/golang/go/issues/27589) to handle this cases at the `encode/json` package, but... so far, I guess that not progress on this one.

Do we think that let it as `null` could cause any issue?

**Comment by rluders** [comment]
> Is the duration above in ms ? It should be obvious from the file IMO. Do we need to have the name with the `github.com/openshift/insights-operator` prefix?

Nope, `time.Duration` will always  use nanoseconds.
Yes, we already use the full name and I guess that it is a good idea to keep it like this, 'cause the package is important to know where the gather lives.

**Comment by tremes** [comment]
@rluders  Well I would divide it and use `elapsed_in_sec` or `elapsed_in_ms`.

**Comment by tremes** [comment]
For the errors the `omitempty` would be better IMO.

File: pkg/record/interface.go [hunk 1]
```diff
@@ -48,12 +48,28 @@ func (m JSONMarshaller) GetExtension() string {
...
 	for _, bulkFn := range bulkFns {
-		klog.V(5).Infof("Gathering %s", runtime.FuncForPC(reflect.ValueOf(bulkFn).Pointer()).Name())
+		gatherName := runtime.FuncForPC(reflect.ValueOf(bulkFn).Pointer()).Name()
+		klog.V(5).Infof("Gathering %s", gatherName)
+
+		start := time.Now()
 		records, errs := bulkFn()
+		elapsed := time.Now().Sub(start).Truncate(time.Millisecond)
+
+		klog.V(4).Infof("Gather %s took %s to process %d records", gatherName, elapsed, len(records))
```
Comment (tremes) [inline pkg/record/interface.go:70]: Why log level 4 here and 5 above? I would do 5 for both. It also seems to me little bit redundant to log the time when we have it in the archive.

Comment (rluders) [reply pkg/record/interface.go:70]: If the recorder fails, you have the logs. Maybe change to 5 is fine, but I was keeping it as the one that measure the execution time of the periodic:

```
klog.V(4).Infof("Periodic gather %s completed in %s", name, time.Now().Sub(start).Truncate(time.Millisecond))
```

Change to 5 or not?

Comment (tremes) [reply pkg/record/interface.go:70]: OK good point. Thanks. Well it's true it's minor but I would be for 5.

**Comment by rluders** [comment]
> @rluders Well I would divide it and use `elapsed_in_sec` or `elapsed_in_ms`.

I don't know. Why both if you can have just one and do the maths? Not that complicated.

**Pull request merged**